File: Day1/Day1.fs
module Day1

    open System 
    
    let input = 
        System.IO.File.ReadLines("Day1/input") 
        |> Seq.map int
    
    let fuelByWeight weight = (weight / 3) - 2

    let rec totalFuel currentTotal fuelWeight =
        if fuelWeight <= 0 then currentTotal
        else totalFuel (currentTotal + fuelWeight) (fuelByWeight fuelWeight)
    
    let Run =
        input
        |> Seq.map fuelByWeight
        |> Seq.sumBy (totalFuel 0)
        |> string
        
// Part 1: 3252897    
// Part 2: 4876469

File: aoc2019rs/src/solutions/day1.rs
use anyhow::Result;


use crate::utils;

pub fn run() -> Result<String> {
    let input = utils::read_input_lines_as_i64(1)?;

    let result = input.into_iter()
        .map(fuel_by_weight)
        .fold(0, total_fuel);
    
    Ok(format!("{}", result))
}

fn fuel_by_weight(weight: i64) -> i64 {
    (weight / 3) - 2
}

fn total_fuel(current_total: i64, weight: i64) -> i64 {
    if weight <= 0 {
        current_total
    } else {
        total_fuel(current_total + weight, fuel_by_weight(weight))
    }
}
